File: termux_aichain/graph/state.py
"""
==============================================================================
termux-aichain Graph Engine: StateGraph & Cyclic Orchestration
==============================================================================
Ultra-lightweight state machine and cyclic multi-agent graph engine.
Zero external heavy dependencies - Pure Python 3.10+ standard library.
"""

from __future__ import annotations
import inspect
import asyncio
from typing import Any, AsyncIterator, Callable, Dict, Iterator, List, Optional, Set, Tuple, Union
from termux_aichain.core.base import Runnable

START = "__start__"
END = "__end__"

class StateGraph:
    """Cyclic state graph orchestrator replacing heavy LangGraph dependencies."""

    def __init__(self, state_schema: Optional[type] = None):
        self.state_schema = state_schema or dict
        self.nodes: Dict[str, Callable[[Dict[str, Any]], Union[Dict[str, Any], Any]]] = {}
        self.edges: Dict[str, str] = {}
        self.conditional_edges: Dict[str, Tuple[Callable[[Dict[str, Any]], str], Optional[Dict[str, str]]]] = {}
        self.entry_point: Optional[str] = None

    def add_node(self, name: str, action: Callable[[Dict[str, Any]], Union[Dict[str, Any], Any]]) -> StateGraph:
        if name in (START, END):
            raise ValueError(f"Cannot name node '{name}': reserved keyword.")
        self.nodes[name] = action
        return self

    def add_edge(self, from_node: str, to_node: str) -> StateGraph:
        if from_node == START:
            self.entry_point = to_node
        else:
            self.edges[from_node] = to_node
        return self

    def add_conditional_edges(
        self,
        source: str,
        router: Callable[[Dict[str, Any]], str],
        path_map: Optional[Dict[str, str]] = None
    ) -> StateGraph:
        self.conditional_edges[source] = (router, path_map)
        return self

    def set_entry_point(self, node_name: str) -> StateGraph:
        self.entry_point = node_name
        return self

    def set_finish_point(self, node_name: str) -> StateGraph:
        self.edges[node_name] = END
        return self

    def compile(self) -> CompiledGraph:
        if not self.entry_point:
            raise ValueError("StateGraph requires an entry point. Call set_entry_point() or add_edge(START, ...).")
        return CompiledGraph(
            nodes=dict(self.nodes),
            edges=dict(self.edges),
            conditional_edges=dict(self.conditional_edges),
            entry_point=self.entry_point,
            state_schema=self.state_schema
        )

class CompiledGraph(Runnable):
    """Executable compiled state graph instance."""

    def __init__(
        self,
        nodes: Dict[str, Callable[[Dict[str, Any]], Any]],
        edges: Dict[str, str],
        conditional_edges: Dict[str, Tuple[Callable[[Dict[str, Any]], str], Optional[Dict[str, str]]]],
        entry_point: str,
        state_schema: type
    ):
        self.nodes = nodes
        self.edges = edges
        self.conditional_edges = conditional_edges
        self.entry_point = entry_point
        self.state_schema = state_schema

    def _get_next_node(self, current_node: str, state: Dict[str, Any]) -> str:
        # Check conditional edge first
        if current_node in self.conditional_edges:
            router, path_map = self.conditional_edges[current_node]
            route_res = router(state)
            if path_map and route_res in path_map:
                return path_map[route_res]
            return route_res
        
        # Check static edge
        return self.edges.get(current_node, END)

    def invoke(self, input_val: Dict[str, Any], max_iterations: int = 50, **kwargs: Any) -> Dict[str, Any]:
        state = dict(input_val) if isinstance(input_val, dict) else {"input": input_val}
        current_node = self.entry_point
        iteration = 0

        while current_node != END and iteration < max_iterations:
            if current_node not in self.nodes:
                raise KeyError(f"Node '{current_node}' not registered in StateGraph.")

            action = self.nodes[current_node]
            update = action(state)
            if isinstance(update, dict):
                state.update(update)

            current_node = self._get_next_node(current_node, state)
            iteration += 1

        if iteration >= max_iterations:
            raise RuntimeError(f"StateGraph exceeded maximum iteration safety limit ({max_iterations}). Possible infinite cycle.")

        return state

    async def ainvoke(self, input_val: Dict[str, Any], max_iterations: int = 50, **kwargs: Any) -> Dict[str, Any]:
        state = dict(input_val) if isinstance(input_val, dict) else {"input": input_val}
        current_node = self.entry_point
        iteration = 0

        while current_node != END and iteration < max_iterations:
            if current_node not in self.nodes:
                raise KeyError(f"Node '{current_node}' not registered in StateGraph.")

            action = self.nodes[current_node]
            if inspect.iscoroutinefunction(action):
                update = await action(state)
            else:
                update = action(state)

            if isinstance(update, dict):
                state.update(update)

            current_node = self._get_next_node(current_node, state)
            iteration += 1

        if iteration >= max_iterations:
            raise RuntimeError(f"StateGraph exceeded maximum iteration safety limit ({max_iterations}).")

        return state

    def stream(self, input_val: Dict[str, Any], max_iterations: int = 50, **kwargs: Any) -> Iterator[Tuple[str, Dict[str, Any]]]:
        state = dict(input_val) if isinstance(input_val, dict) else {"input": input_val}
        current_node = self.entry_point
        iteration = 0

        while current_node != END and iteration < max_iterations:
            if current_node not in self.nodes:
                raise KeyError(f"Node '{current_node}' not registered in StateGraph.")

            action = self.nodes[current_node]
            update = action(state)
            if isinstance(update, dict):
                state.update(update)

            yield (current_node, dict(state))

            current_node = self._get_next_node(current_node, state)
            iteration += 1

    async def astream(self, input_val: Dict[str, Any], max_iterations: int = 50, **kwargs: Any) -> AsyncIterator[Tuple[str, Dict[str, Any]]]:
        state = dict(input_val) if isinstance(input_val, dict) else {"input": input_val}
        current_node = self.entry_point
        iteration = 0

        while current_node != END and iteration < max_iterations:
            if current_node not in self.nodes:
                raise KeyError(f"Node '{current_node}' not registered in StateGraph.")

            action = self.nodes[current_node]
            if inspect.iscoroutinefunction(action):
                update = await action(state)
            else:
                update = action(state)

            if isinstance(update, dict):
                state.update(update)

            yield (current_node, dict(state))

            current_node = self._get_next_node(current_node, state)
            iteration += 1

    def __repr__(self) -> str:
        return f"CompiledGraph(nodes={list(self.nodes.keys())}, entry_point='{self.entry_point}')"

File: js/esm/graph/state.js
/**
 * ==============================================================================
 * @termux-ai/chain Graph Engine: StateGraph (TypeScript ESM)
 * ==============================================================================
 */
import { createPipeline } from "../core/base.js";
export const START = "__start__";
export const END = "__end__";
export class StateGraph {
    nodes = new Map();
    edges = new Map();
    conditionalEdges = new Map();
    entryPoint;
    addNode(name, action) {
        if (name === START || name === END) {
            throw new Error(`Cannot use reserved name '${name}' as node.`);
        }
        this.nodes.set(name, action);
        return this;
    }
    addEdge(fromNode, toNode) {
        if (fromNode === START) {
            this.entryPoint = toNode;
        }
        else {
            this.edges.set(fromNode, toNode);
        }
        return this;
    }
    addConditionalEdges(source, router, pathMap) {
        this.conditionalEdges.set(source, { router, pathMap });
        return this;
    }
    setEntryPoint(nodeName) {
        this.entryPoint = nodeName;
        return this;
    }
    setFinishPoint(nodeName) {
        this.edges.set(nodeName, END);
        return this;
    }
    compile() {
        if (!this.entryPoint) {
            throw new Error("StateGraph requires an entry point. Use setEntryPoint() or addEdge(START, ...).");
        }
        return new CompiledGraph(new Map(this.nodes), new Map(this.edges), new Map(this.conditionalEdges), this.entryPoint);
    }
}
export class CompiledGraph {
    nodes;
    edges;
    conditionalEdges;
    entryPoint;
    constructor(nodes, edges, conditionalEdges, entryPoint) {
        this.nodes = nodes;
        this.edges = edges;
        this.conditionalEdges = conditionalEdges;
        this.entryPoint = entryPoint;
    }
    getNextNode(currentNode, state) {
        const cond = this.conditionalEdges.get(currentNode);
        if (cond) {
            const res = cond.router(state);
            if (cond.pathMap && res in cond.pathMap) {
                return cond.pathMap[res];
            }
            return res;
        }
        return this.edges.get(currentNode) ?? END;
    }
    async invoke(input, options) {
        const state = { ...input };
        let currentNode = this.entryPoint;
        const maxIterations = options?.maxIterations ?? 50;
        let iter = 0;
        while (currentNode !== END && iter < maxIterations) {
            const action = this.nodes.get(currentNode);
            if (!action) {
                throw new Error(`Node '${currentNode}' is not defined in graph.`);
            }
            const update = await action(state);
            if (update && typeof update === "object") {
                Object.assign(state, update);
            }
            currentNode = this.getNextNode(currentNode, state);
            iter++;
        }
        if (iter >= maxIterations) {
            throw new Error(`StateGraph exceeded maxIterations limit (${maxIterations}).`);
        }
        return state;
    }
    async *stream(input, options) {
        const state = { ...input };
        let currentNode = this.entryPoint;
        const maxIterations = options?.maxIterations ?? 50;
        let iter = 0;
        while (currentNode !== END && iter < maxIterations) {
            const action = this.nodes.get(currentNode);
            if (!action) {
                throw new Error(`Node '${currentNode}' is not defined in graph.`);
            }
            const update = await action(state);
            if (update && typeof update === "object") {
                Object.assign(state, update);
            }
            yield [currentNode, { ...state }];
            currentNode = this.getNextNode(currentNode, state);
            iter++;
        }
    }
    pipe(next) {
        return createPipeline([this, next]);
    }
}
